docs/assets/diagrams/transfer/06-transaction-feedback.puml

```diff
@@ -0,0 +1,65 @@
+@startuml
+
+title Transaction Feedback
+
+hide footbox
+
+autonumber
+
+box "Mobile Device" #Light
+  actor Alice
+  participant "PISP App" as App
+end box
+box "PISP" #Light
+  database Firebase
+  participant "PISP Server" as Server
+end box
+box "Mojaloop" #Light
+  participant "Switch" as Mojaloop
+end box
+
+activate Mojaloop
+
+rnote left of Mojaloop #Light
+  ""**PUT /thirdpartyRequests/transactions/456**""
```

Comment:
This will be `PUT /transfers/{id}` now, based on more recent design discussions, as it is an exact forwarded request from the PayeeFSP to the switch

Reply:
okay, can I check what's the format of the callback? is the current one still correct?

Reply:
@lewisdaly based on the [API Definition v1.0](https://docs.mojaloop.io/mojaloop-specification/documents/API%20Definition%20v1.0.html#6731-put-transfersid), `PUT /transfers/{ID}` does not contain any information about the `transactionRequestId`, which is needed to let PISP identify the transaction, any idea on how we could get that information?

Reply:
Yeah, well spotted.

So there's a couple ideas floating around for this:
1. The `transactionRequestId` is encoded in the `ilpPacket`, so worst case scenario, we can decode the `ilpPacket`
2. Mojaloop API v2.0 looks likely to include the `transactionRequestId` in the request body
3. We're also thinking that we can let the PISP specify the `transferId` ahead of time during the `POST thirdpartyRequests/transaction`

Reply:
> 1. The `transactionRequestId` is encoded in the `ilpPacket`, so worst case scenario, we can decode the `ilpPacket`

But `PUT /transfers/{ID}` also does not contain `ilpPacket` right? Or perhaps is it included inside the `fulfilment`?

> 2. Mojaloop API v2.0 looks likely to include the `transactionRequestId` in the request body
> 3. We're also thinking that we can let the PISP specify the `transferId` ahead of time during the `POST thirdpartyRequests/transaction`

If we are going to include both `transactionId` and `transactionRequestId` everywhere, may I ask about why do we need to separate those values? Wouldn't it be better to have a single id that identifies the whole transaction process?

Reply:
>But PUT /transfers/{ID} also does not contain ilpPacket right? Or perhaps is it included inside the fulfilment?

Yeah you're right, I was getting confused with `POST /transfers`

>If we are going to include both transactionId and transactionRequestId everywhere, may I ask about why do we need to separate those values? Wouldn't it be better to have a single id that identifies the whole transaction process?

Sorry I meant just `transactionId`, not `transactionRequestId`.

Reply:
> Sorry I meant just `transactionId`, not `transactionRequestId`.

Sorry, could you please reiterate on this? If I understand it correctly, PISP needs the `transactionRequestId` to identify the transfer right? And unless it is the one who decides on the `transactionId`, I assume that getting the value won't be useful? However, my concern here is that, if PISP is the one that decides on both values, why do we need to differentiate them?

Reply:
Yeah no worries. You've uncovered a part of the design that has more _hand waving_ than the rest, and we aren't 100% solid on what this will look like just yet.

I think we're leaning on the idea that a PISP can specify the `transferId` and `transactionId` ahead of time in the `POST /thirdpartyRequests/transactions` request. That way, it _can_ understand and identify incoming `PUT /transfers/{id}` requests, since it was one who specified the `transferId`.

> However, my concern here is that, if PISP is the one that decides on both values, why do we need to differentiate them? 

You're right - I guess we don't strictly need to differentiate them, but remember that they have different purposes. A transaction (and hence a `transactionId`) can live outside the scope of a transaction request. 

If `transactionId` is added to the `Transfer` object, it would be a _required_ field (I think), because a transfer must always be within a bigger transaction (even if the transaction only contains 1 transfer).

A `TransactionRequest` (and for our purposes, `ThirdpartyTransactionRequest`) may not result in the creation of a `Transfer` or `Transaction`.

Additionally, not every `Transfer` is the result of a `TransactionRequest`, so if we added the `transactionRequestId` to the `Transfer` Object (which I don't think is being discussed at the moment, but I could be wrong), it would be optional, much like in the `Quotes` Object. (see [post-quotes](https://docs.mojaloop.io/mojaloop-specification/documents/API%20Definition%20v1.0.html#6522-post-quotes))

Reply:
Ahh I see, okay got it, thanks for the robust explanation!

Besides, I just realised something with the current transfer flow design...

If we go with that option, the database for PISP will need to have an additional index, which is for the `transactionId`, because it needs to handle `POST /authorizations`  with only `transactionRequestId` being provided, but also need to handle `PUT /transfers/{ID}` with only `transactionId` being provided.

Is it safe to assume that we can set `transactionId` equal to `transactionRequestId` to simplify the indexing?

Reply:
After looking back at the sequence diagram, I just realized that `POST /authorizations` actually contains both `transactionRequestId` and `transactionId`, which essentially solves the issue that we discussed here 😂 ... Hence, it's fine to let DFSP to decide on the `transactionId`. The PISP can create a mapping between `transactionRequestId` and `transactionId` upon receiving `POST /authorizations`.

In the current design, I still set `transactionRequestId` as a document ID in Firebase to ensure uniqueness, and just believe that `transactionId` will be unique as well... but in case `transactionId` is not unique, we will just end up updating multiple transaction documents and I think its beyond the control of the PISP...

Reply:
Ah ok. I forgot about how cloud firestore works in relation to indicies.

>Hence, it's fine to let DFSP to decide on the transactionId. The PISP can create a mapping between transactionRequestId and transactionId upon receiving POST /authorizations.

Sure, as long as you can build that relationship, I think we're fine, and can proceed for now. But I do think that the PISP will be setting the `transactionId`. And we _can_ assume a `transactionId` is unique.

> Besides, @lewisdaly, can I check whether the id in PUT /transfers/{ID} is equal to transactionId? Or perhaps there could be a different transferId? For now, I assume that they are different, but there's a transactionId in the request body.

Yeah I think we underestimated the challenges for Mojaloop Clients with respect to this callback. All I can say for now is watch this space as we still have unresolved design issues here. For now however, maybe we can assume that the `PUT /transfers/{ID}` will _contain_ a `transactionId` in the body (which it doesn't at the moment).